go/networkqualityd.go

```diff
@@ -91,17 +100,63 @@ func main() {
 		}()
 	}
 
+	if *announce {
+		ips := make([]net.IP, 0)
+		// The user may give us a hostname (rather than an address to listen on). In order to
+		// handle this situation, we will use DNS to convert it to an IP. As a result, we may
+		// get back more than one address -- handle that!
+		if addresses, lookupErr := net.LookupHost(*listenAddr); lookupErr == nil {
+			for _, addr := range addresses {
+				if parsedAddr := net.ParseIP(addr); parsedAddr != nil {
+					ips = append(ips, parsedAddr)
+				}
+			}
+		}
+		if announceResponder, announceHandle, announceErr := configureAnnouncer(ips, *configName, *configPort); announceErr == nil {
+			defer announceResponder.Remove(announceHandle)
+			go announceResponder.Respond(operatingCtx)
+		} else {
+			log.Printf("Warning: Could not announce the server instance: %v.\n", announceErr)
+		}
+	}
+
 	var wg sync.WaitGroup
 	wg.Add(1)
 
-	listenAddr := fmt.Sprintf("%s:%d", *listenAddr, *configPort)
-	go func(listenAddr string) {
+	server := &http.Server{}
+	server.Addr = fmt.Sprintf("%s:%d", *listenAddr, *configPort)
+	server.Handler = mux
+
+	// Create an http2.Server simply for the purposes of configuring
+	// certain performance-related parameters. Use the
+	// http2.ConfigureServer function to merge the HTTP/2-specific
+	// configuration parameters in to the (generic) HTTP server.
+	http2Configuration := &http2.Server{}
+	http2Configuration.MaxUploadBufferPerConnection = 1 << 27
+	http2Configuration.MaxUploadBufferPerStream = 1 << 27
+	http2.ConfigureServer(server, http2Configuration)
+
+	go func() {
 		log.Printf("Network Quality URL: https://%s:%d%s/config", *configName, *configPort, *contextPath)
-		if err := http.ListenAndServeTLS(listenAddr, *certFilename, *keyFilename, mux); err != nil {
+		if err := server.ListenAndServeTLS(*certFilename, *keyFilename); err != nil && err != http.ErrServerClosed {
 			log.Fatal(err)
 		}
 		wg.Done()
-	}(listenAddr)
+	}()
+
+	// The user can stop the server with SIGINT
+	signalChannel := make(chan os.Signal, 1)   // make the channel buffered, per documentation.
+	signal.Notify(signalChannel, os.Interrupt) // only Interrupt is guaranteed to exist on all platforms.
+
+SignalLoop:
```

Comment:
Curious, is it expected this now delays shutdown?

Reply:
It is my understanding (and confirmed in testing) that this should *not* make shutdown delayed. In fact, if you send it a `SIGKILL` (or do a Ctrl-C) then this handler will not fire and the program exits immediately. Obviously not what we want when `--announce` is in effect, but those are the breaks! Let me know if this response is helpful?

Reply:
Let me stand corrected: it does appear that `os.Interrupt` catches Ctrl-C! In my opinion, that is good news.

Reply:
The signal handler aspect works. Debugging it looks like there are 2 issues. The loop is an infinite loop and takes time for `announceResponder.Remove` to do its thing (multiple seconds). I'll fix this in another PR.

Reply:
The shutdown delay is probably a macOS thing. The `unannounce` in dnssd has a 250ms sleep. On my Mac I think it's unannouncing over 18 interfaces(vpns, other virtual interfaces, etc) with that sleep in-between each interface.